[CORE-13370] archival: Fix retention fallthrough behavior

Lazin opened the pull request:
In the `ntp_archiver::apply_retention` method we're not throwing exception if we failed to replicate the command that moves start offset forward. This leads to a situation when even if we failed to replicate the command the `apply_gc` and `apply_spillover` methods are invoked. This leads to a situation when spillover invariant is broken. The `apply_spillover` method expects that the start offset is aligned with the first segment in the STM manifest. If this is not the case the spillover manifest that it produces can't be applied.

This is what leads to this problem:
- `apply_retention` fails to replicate the command due to timeout. The timeout could be caused by the high cpu load. The command is still replciated and applied asynchronously.
- `apply_gc` invoked before the command that moves start offset is applied. The method does nothing since start offset is still aligned with the first segment in the manifest.
- `apply_spillover` is invoked after the command is applied. Now the start offset is moved forward and is no longer aligned with the first segment in the manifest. The method creates a spillover manifest, uploads it, puts it to the cloud storage cache, and replicates the command that adds spill metadata to the manifest. The command fails because it can't pass the check in the manifest (the manifest expects that the base offset of the spill is equal to start offset.

Before https://github.com/redpanda-data/redpanda/pull/27714 was merged this could cause cloud storage cache to fill up. Now this can only create a single failed spillover attempt. 

<!--
See https://github.com/redpanda-data/redpanda/blob/dev/CONTRIBUTING.md#pull-request-body
for more details and examples of what is expected in a PR body.

Content in this top section is REQUIRED. Describe, in plain language, the motivation
behind the change (bug fix, feature, improvement) in this PR and how the included
commits address it.

Add the GitHub keyword `Fixes` to link to bug(s) this PR will fix, e.g.
  Fixes #ISSUE-NUMBER, Fixes #ISSUE-NUMBER, ...

If this PR is a backport, link to the original with `Backport of PR`, e.g.
  Backport of PR #PR-NUMBER
-->

## Backports Required

<!-- Checking at least one of the checkboxes is REQUIRED if this PR is not a backport. -->

- [ ] none - not a bug fix
- [ ] none - this is a backport
- [ ] none - issue does not exist in previous branches
- [ ] none - papercut/not impactful enough to backport
- [x] v25.2.x
- [x] v25.1.x
- [ ] v24.3.x

## Release Notes

<!--
If the changes in this PR do not need to be mentioned in the release
notes, then don't add a sub-section and simply list `none`, e.g.

* none

Otherwise, adding a sub-section or `none` is REQUIRED if the PR is not a backport PR.
If this is a backport PR, adding contents to this section will override
the release notes section inherited from the original PR to dev.

Add one or more of the sub-sections with a short description bullet
point of the change, e.g.

### Bug Fixes

* Short description of the bug fix if this is a PR to `dev` branch.

### Features

* Short description of the feature. Explain how to configure.

### Improvements

* Short description of how this PR improves existing behavior.

-->
* none

Lazin committed:
archival: Enforce spillover invariant

The 'apply_spillover' method expects that the start offset field in the
manifest is aligned with the first segment in the manifest.
This commit explicitly checks this invariant.

Signed-off-by: Evgeny Lazin <4lazin@gmail.com>

Lazin committed:
archival: If throw in apply_retnetion

If apply_retention method failed to replicate the command it's not safe
to run GC or spillover because the command could be actually replicated
and applied asynchronously.

Signed-off-by: Evgeny Lazin <4lazin@gmail.com>

vbotbuildovich commented:
<!--FAILURES SLASH COMMAND https://buildkite.com/redpanda/redpanda/builds/73642-->
<!--adb97914-8945-41fd-9ac6-e2f230fe2f2a-->
<!--90776ece-6a67-4c1b-b3dd-ca671e215024-->
<!--3a6bfef8-c1c6-4417-afce-54ba877614ff-->
<!--25597071-7a8b-4605-ac10-8377a46d1c22-->
<!--title-->
## Retry command for [Build#73642](https://buildkite.com/redpanda/redpanda/builds/73642)
please wait until all jobs are finished before running the slash command
<!--/title-->
<!--body-->
```
/ci-repeat 1
tests/rptest/tests/cloud_storage_scrubber_test.py::CloudStorageScrubberTest.test_scrubber@{"cloud_storage_type":2}
tests/rptest/tests/cloud_storage_scrubber_test.py::CloudStorageScrubberTest.test_scrubber@{"cloud_storage_type":1}
```
<!--/body-->

vbotbuildovich commented:
<!--TEST RESULTS-->
<!--7cbce687-6552-4ee9-952b-1f7a020246c5-->
<!--TEST RESULTS-->
<!--436a6bde-4192-4446-8171-aac6bfd864db-->
<!--title-->
## CI test results
<!--/title-->
<!--body-->
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/73642">build#73642</a></summary>


| test_class                          | test_method                                | test_arguments                                                     | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                  | test_history                                                                                                                                                               |
|:------------------------------------|:-------------------------------------------|:-------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:------------------------------------------------------------------------------------------------------------------------------------------------------------------------|:---------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| CloudStorageScrubberTest            | test_scrubber                              | {"cloud_storage_type": 2}                                          | integration | https://buildkite.com/redpanda/redpanda/builds/73642#0199ba8a-ba60-4d71-9a40-7bbd52eeea28 | FLAKY         | 2/21     | upstream reliability is '100.0'. current run reliability is '9.523809523809524'. drift is 90.47619 and the allowed drift is set to 50. The test should FAIL             | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=CloudStorageScrubberTest&test_method=test_scrubber                       |
| CloudStorageScrubberTest            | test_scrubber                              | {"cloud_storage_type": 2}                                          | integration | https://buildkite.com/redpanda/redpanda/builds/73642#0199ba8b-297a-4401-bd5d-c9ccc187a024 | FLAKY         | 1/21     | upstream reliability is '100.0'. current run reliability is '4.761904761904762'. drift is 95.2381 and the allowed drift is set to 50. The test should FAIL              | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=CloudStorageScrubberTest&test_method=test_scrubber                       |
| CloudStorageScrubberTest            | test_scrubber                              | {"cloud_storage_type": 1}                                          | integration | https://buildkite.com/redpanda/redpanda/builds/73642#0199ba8a-ba62-4431-8570-1e6951b29a86 | FLAKY         | 2/21     | upstream reliability is '100.0'. current run reliability is '9.523809523809524'. drift is 90.47619 and the allowed drift is set to 50. The test should FAIL             | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=CloudStorageScrubberTest&test_method=test_scrubber                       |
| CloudStorageScrubberTest            | test_scrubber                              | {"cloud_storage_type": 1}                                          | integration | https://buildkite.com/redpanda/redpanda/builds/73642#0199ba8b-297b-431e-b360-a363bf3fb157 | FLAKY         | 1/21     | upstream reliability is '100.0'. current run reliability is '4.761904761904762'. drift is 95.2381 and the allowed drift is set to 50. The test should FAIL              | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=CloudStorageScrubberTest&test_method=test_scrubber                       |
| DeleteRecordsTest                   | test_delete_records_concurrent_truncations | {"cloud_storage_enabled": true, "truncate_point": "random_offset"} | integration | https://buildkite.com/redpanda/redpanda/builds/73642#0199ba8a-ba6e-4ed7-bc6a-ed771284140e | FLAKY         | 15/21    | upstream reliability is '100.0'. current run reliability is '71.42857142857143'. drift is 28.57143 and the allowed drift is set to 50. The test should PASS             | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=DeleteRecordsTest&test_method=test_delete_records_concurrent_truncations |
| ShardPlacementTest                  | test_core_count_change                     | null                                                               | integration | https://buildkite.com/redpanda/redpanda/builds/73642#0199ba8b-297b-431e-b360-a363bf3fb157 | FLAKY         | 20/21    | upstream reliability is '100.0'. current run reliability is '95.23809523809523'. drift is 4.7619 and the allowed drift is set to 50. The test should PASS               | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=ShardPlacementTest&test_method=test_core_count_change                    |
| WriteCachingFailureInjectionE2ETest | test_crash_all                             | {"use_transactions": false}                                        | integration | https://buildkite.com/redpanda/redpanda/builds/73642#0199ba8b-2980-4598-9dbd-4aad69c91cec | FLAKY         | 19/21    | upstream reliability is '83.9544513457557'. current run reliability is '90.47619047619048'. drift is -6.52174 and the allowed drift is set to 50. The test should PASS  | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=WriteCachingFailureInjectionE2ETest&test_method=test_crash_all           |
| WriteCachingFailureInjectionTest    | test_unavoidable_data_loss                 | null                                                               | integration | https://buildkite.com/redpanda/redpanda/builds/73642#0199ba8a-ba67-4b9f-b722-7d7a33ef482d | FLAKY         | 20/21    | upstream reliability is '92.26277372262773'. current run reliability is '95.23809523809523'. drift is -2.97532 and the allowed drift is set to 50. The test should PASS | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=WriteCachingFailureInjectionTest&test_method=test_unavoidable_data_loss  |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/75060">build#75060</a></summary>


| test_class                        | test_method                                | test_arguments                                                                | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                  | test_history                                                                                                                                                                               |
|:----------------------------------|:-------------------------------------------|:------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:------------------------------------------------------------------------------------------------------------------------------------------------------------------------|:-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| CloudStorageScrubberTest          | test_scrubber                              | {"cloud_storage_type": 2}                                                     | integration | https://buildkite.com/redpanda/redpanda/builds/75060#019a2727-6076-46b9-91db-38f8419b10c5 | FLAKY         | 5/21     | upstream reliability is '100.0'. current run reliability is '23.809523809523807'. drift is 76.19048 and the allowed drift is set to 50. The test should FAIL            | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=CloudStorageScrubberTest&test_method=test_scrubber                                       |
| CloudStorageScrubberTest          | test_scrubber                              | {"cloud_storage_type": 2}                                                     | integration | https://buildkite.com/redpanda/redpanda/builds/75060#019a2727-f278-4262-be10-ae052b508a2e | FLAKY         | 5/21     | upstream reliability is '100.0'. current run reliability is '23.809523809523807'. drift is 76.19048 and the allowed drift is set to 50. The test should FAIL            | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=CloudStorageScrubberTest&test_method=test_scrubber                                       |
| CloudStorageScrubberTest          | test_scrubber                              | {"cloud_storage_type": 1}                                                     | integration | https://buildkite.com/redpanda/redpanda/builds/75060#019a2727-6077-44c5-a5b7-3b0ff5f327d9 | FLAKY         | 2/21     | upstream reliability is '100.0'. current run reliability is '9.523809523809524'. drift is 90.47619 and the allowed drift is set to 50. The test should FAIL             | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=CloudStorageScrubberTest&test_method=test_scrubber                                       |
| CloudStorageScrubberTest          | test_scrubber                              | {"cloud_storage_type": 1}                                                     | integration | https://buildkite.com/redpanda/redpanda/builds/75060#019a2727-f279-42f6-b40f-c07110c49471 | FLAKY         | 2/21     | upstream reliability is '100.0'. current run reliability is '9.523809523809524'. drift is 90.47619 and the allowed drift is set to 50. The test should FAIL             | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=CloudStorageScrubberTest&test_method=test_scrubber                                       |
| MountUnmountIcebergTest           | test_simple_remount                        | {"cloud_storage_type": 1}                                                     | integration | https://buildkite.com/redpanda/redpanda/builds/75060#019a2727-607e-4269-84eb-25c38b7bc5c0 | FLAKY         | 15/21    | upstream reliability is '84.85523385300668'. current run reliability is '71.42857142857143'. drift is 13.42666 and the allowed drift is set to 50. The test should PASS | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=MountUnmountIcebergTest&test_method=test_simple_remount                                  |
| DeleteRecordsTest                 | test_delete_records_concurrent_truncations | {"cloud_storage_enabled": true, "truncate_point": "one_below_high_watermark"} | integration | https://buildkite.com/redpanda/redpanda/builds/75060#019a2727-f27e-4bec-b2df-2fa875947fc5 | FLAKY         | 15/21    | upstream reliability is '100.0'. current run reliability is '71.42857142857143'. drift is 28.57143 and the allowed drift is set to 50. The test should PASS             | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=DeleteRecordsTest&test_method=test_delete_records_concurrent_truncations                 |
| LogCompactionTxRemovalUpgradeTest | test_tx_control_batch_removal_with_upgrade | {"test_case_name": "All aborts"}                                              | integration | https://buildkite.com/redpanda/redpanda/builds/75060#019a2727-f28e-4741-9274-0203a7e794f3 | FLAKY         | 16/21    | upstream reliability is '100.0'. current run reliability is '76.19047619047619'. drift is 23.80952 and the allowed drift is set to 50. The test should PASS             | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=LogCompactionTxRemovalUpgradeTest&test_method=test_tx_control_batch_removal_with_upgrade |
| PartitionMovementTest             | test_static                                | {"num_to_upgrade": 0}                                                         | integration | https://buildkite.com/redpanda/redpanda/builds/75060#019a2727-607d-4489-ace0-6d48d1e7b26a | FLAKY         | 15/21    | upstream reliability is '88.66442199775533'. current run reliability is '71.42857142857143'. drift is 17.23585 and the allowed drift is set to 50. The test should PASS | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=PartitionMovementTest&test_method=test_static                                            |
| TLSVersionTestRSA                 | test_change_version                        | {"version": 1}                                                                | integration | https://buildkite.com/redpanda/redpanda/builds/75060#019a2727-6078-4d5d-9e65-a17104ef22e8 | FAIL          | 0/1      |                                                                                                                                                                         | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=TLSVersionTestRSA&test_method=test_change_version                                        |
</details>
<!--/body-->

oleiman commented on src/v/cluster/archival/ntp_archiver_service.cc:
```diff
@@ -2994,6 +2994,23 @@ ss::future<> ntp_archiver::apply_spillover() {
+        auto fo = manifest().begin()->base_offset;
+        if (fo != so.value()) {
+            vlog(
+              _rtclog.error,
```
probably want this at WARN 😄 

https://ci-artifacts.dev.vectorized.cloud/redpanda/73642/0199ba8a-ba60-4d71-9a40-7bbd52eeea28/vbuild/ducktape/results/final/report.html

Lazin replied:
that should be fixed, the scrubber test triggered it because it resets the manifest to introduce the inconsistency

oleiman reviewed (commented):
lgtm modulo log level 😄 

one piece of context that's missing from the description is that the backgrounded replicate command holds an exclusive lock over the metadata STM lock as well, so short-circuiting `housekeeping` in this way guarantees that we won't call gc/spillover until the truncate command is replicated. Is that right? sorry if obvious.

andrwng commented on src/v/cluster/archival/ntp_archiver_service.cc:
```diff
@@ -2994,6 +2994,23 @@ ss::future<> ntp_archiver::apply_spillover() {
     const auto manifest_upload_timeout = _conf->manifest_upload_timeout();
     const auto manifest_upload_backoff = _conf->cloud_storage_initial_backoff();
 
+    // Check the spillover invariant.
```
Why isn't the fix to demote the archival_metadata_stm error to a warning? The issue is benign at this point, and expected due to races, right?

Lazin replied:
Isn't it better to eliminate the race?

WillemKauf commented:
/ci-repeat 1

vbotbuildovich commented:
<!--FAILURES SLASH COMMAND https://buildkite.com/redpanda/redpanda/builds/75060-->
<!--c596621d-6f94-490f-97e9-a7f61791938a-->
<!--f972ba07-2cfe-49e3-9855-e7b765209ffd-->
<!--66264dc5-4049-4c50-9ad6-40307148d3ac-->
<!--b9aa8277-87d2-4150-83b8-c6d53921463e-->
<!--86bea0ff-e410-4710-9b53-97160a0529eb-->
<!--title-->
## Retry command for [Build#75060](https://buildkite.com/redpanda/redpanda/builds/75060)
please wait until all jobs are finished before running the slash command
<!--/title-->
<!--body-->
```
/ci-repeat 1
tests/rptest/tests/tls_version_test.py::TLSVersionTestRSA.test_change_version@{"version":1}
tests/rptest/tests/cloud_storage_scrubber_test.py::CloudStorageScrubberTest.test_scrubber@{"cloud_storage_type":1}
tests/rptest/tests/cloud_storage_scrubber_test.py::CloudStorageScrubberTest.test_scrubber@{"cloud_storage_type":2}
```
<!--/body-->

Lazin committed:
rptest: Allow spillover violation error message

in scrubber test because the test explicitly violates consistency.

Signed-off-by: Evgeny Lazin <4lazin@gmail.com>

Lazin committed:
ts: Fix ODR violation

The recovery subsystem in cloud_storage defines error_category function
with exactly the same signature etc. This commit renames it so there is
no name collision.

Signed-off-by: Evgeny Lazin <4lazin@gmail.com>

oleiman commented on tests/rptest/tests/cloud_storage_scrubber_test.py:
```diff
@@ -54,6 +54,10 @@
     # manifest will refuse to apply the command and log the error below.
     r"cloud_storage - .* New replacement segment does not line up with previous segment",
     r"cluster - .* Can't add segment:",
+    # The same could happen with spillover. Spillover assumes that the start offset is
+    # aligned with the first segment in the manifest but this may not be the case after
+    # after the manifest reset.
+    r"archival - .* Spillover invariant violated",
```
👍

oleiman commented on src/v/cloud_storage/recovery_errors.cc:
sorry, how does this relate to the PR?

oleiman reviewed (approved):
lgtm. what's the deal with the last commit?

Lazin commented:
> lgtm. what's the deal with the last commit?

it's unrelated, I stumbled on this when I was fixing the CI failure

vbotbuildovich commented:
/backport v25.2.x

vbotbuildovich commented:
/backport v25.1.x

dotnwat commented on src/v/cloud_storage/recovery_errors.cc:
```diff
@@ -37,13 +37,13 @@ std::string recovery_error_category::message(int c) const {
 
-const std::error_category& error_category() noexcept {
-    static const recovery_error_category e;
+const std::error_category& recovery_error_category() noexcept {
+    static const recovery_error_category_type e;
```
@Lazin maybe we can do something like this in cloud topics?
